File: PrimaMateria/PrimaMateria/Classes/XTRElementTableViewCell.m
//
//  XTRElementTableViewCell.m
//  PrimaMateria
//
//  Created by Jerry Porter on 4/8/16.
//  Copyright © 2016 xTrensa. All rights reserved.
//

@interface XTRElementTableViewCell()

-(void)colorChanged: (NSNotification *)aNotification;

@end

@implementation XTRElementTableViewCell

-(void)colorChanged: (NSNotification *)aNotification {
    [self setupColors];
    [self setNeedsDisplay];
}

- (void)setupColors {
    UIColor *aSeriesColor = self.element.seriesColor;
    self.atomicNumberLabel.backgroundColor = aSeriesColor;
    self.symbolLabel.backgroundColor = aSeriesColor;
    self.nameLabel.backgroundColor = aSeriesColor;
    self.atomicMassLabel.backgroundColor = aSeriesColor;
    self.boilingPointLabel.backgroundColor = aSeriesColor;
    self.meltingPointLabel.backgroundColor = aSeriesColor;
    self.densityLabel.backgroundColor = aSeriesColor;
    self.seriesLabel.backgroundColor = aSeriesColor;
    self.periodLabel.backgroundColor = aSeriesColor;
    self.groupLabel.backgroundColor = aSeriesColor;
}

- (instancetype)initWithStyle: (UITableViewCellStyle)style reuseIdentifier: (NSString *)reuseIdentifier andElement: (XTRElement *) anElement {
    if (self = [super initWithStyle:(UITableViewCellStyle) style reuseIdentifier:reuseIdentifier]) {
        self.element = anElement;
        self.backgroundColor = UIColor.blackColor;
        self.contentView.backgroundColor = UIColor.darkGrayColor;

        self.atomicNumberLabel = [[DynoTableLabel alloc] initWithFrame: CGRectMake(0, 0, 85, 42) andColor: self.element.seriesColor];
        [self.contentView addSubview: self.atomicNumberLabel];

        self.symbolLabel = [[DynoTableLabel alloc] initWithFrame: CGRectMake(86, 0, 95, 42) andColor: self.element.seriesColor];
        [self.contentView addSubview: self.symbolLabel];

        self.nameLabel = [[DynoTableLabel alloc] initWithFrame: CGRectMake(182, 0, 125, 42) andColor: self.element.seriesColor];
        [self.contentView addSubview: self.nameLabel];

        self.atomicMassLabel = [[DynoTableLabel alloc] initWithFrame: CGRectMake(308, 0, 95, 42) andColor: self.element.seriesColor];
        [self.contentView addSubview: self.atomicMassLabel];

        self.boilingPointLabel = [[DynoTableLabel alloc] initWithFrame: CGRectMake(404, 0, 105, 42) andColor: self.element.seriesColor];
        [self.contentView addSubview: self.boilingPointLabel];

        self.meltingPointLabel = [[DynoTableLabel alloc] initWithFrame: CGRectMake(510, 0, 105, 42) andColor: self.element.seriesColor];
        [self.contentView addSubview: self.meltingPointLabel];

        self.densityLabel = [[DynoTableLabel alloc] initWithFrame: CGRectMake(616, 0, 105, 42) andColor: self.element.seriesColor];
        [self.contentView addSubview: self.densityLabel];

        self.seriesLabel = [[DynoTableLabel alloc] initWithFrame: CGRectMake(722, 0, 160, 42) andColor: self.element.seriesColor];
        [self.contentView addSubview: self.seriesLabel];

        self.periodLabel = [[DynoTableLabel alloc] initWithFrame: CGRectMake(883, 0, 72, 42) andColor: self.element.seriesColor];
        [self.contentView addSubview: self.periodLabel];

        self.groupLabel = [[DynoTableLabel alloc] initWithFrame: CGRectMake(956, 0, 68, 42) andColor: self.element.seriesColor];
        [self.contentView addSubview: self.groupLabel];
        
        [[NSNotificationCenter defaultCenter] addObserver: self selector: @selector(colorChanged:) name: NOTIFICATION_SERIES_COLOR_CHANGED object: nil];
    }
    return self;
}

- (void)modifyLabel: (DynoTableLabel *)aLabel withStandardConditionColor: (UIColor *)aStandardConditionColor seriesColor: (UIColor *)aSeriesColor {
    aLabel.font = [UIFont systemFontOfSize: 16.0];
    aLabel.textColor = aStandardConditionColor;
    aLabel.backgroundColor = aSeriesColor;
    aLabel.autoresizingMask = UIViewAutoresizingFlexibleRightMargin | UIViewAutoresizingFlexibleHeight;
}

- (void)modifyCellProperties {
    UIColor *seriesColor = self.element.seriesColor;
    UIColor *selectedColor = [UIColor reverseColorOf: seriesColor];
    self.selectedBackgroundView = [[UIImageView alloc] initWithImage: [UIImage imageWithColor: selectedColor size:self.frame.size]];

    [self modifyLabel: self.atomicNumberLabel withStandardConditionColor: self.element.standardConditionColor seriesColor: self.element.seriesColor];
    self.atomicNumberLabel.text = [NSString stringWithFormat: @"%@", self.element.atomicNumber];
    self.atomicNumberLabel.textAlignment = NSTextAlignmentRight;

    [self modifyLabel: self.symbolLabel withStandardConditionColor: self.element.standardConditionColor seriesColor: self.element.seriesColor];
    self.symbolLabel.font = [UIFont systemFontOfSize:26.0];
    self.symbolLabel.text = [NSString stringWithFormat: @"%@", self.element.symbol];
    self.symbolLabel.textAlignment = NSTextAlignmentCenter;

    [self modifyLabel: self.nameLabel withStandardConditionColor: self.element.standardConditionColor seriesColor: self.element.seriesColor];
    self.nameLabel.text = [NSString stringWithFormat: @"%@", self.element.name];
    self.nameLabel.textAlignment = NSTextAlignmentLeft;

    [self modifyLabel: self.atomicMassLabel withStandardConditionColor: self.element.standardConditionColor seriesColor: self.element.seriesColor];
    self.atomicMassLabel.text = [NSString stringWithFormat: @"%@", self.element.atomicMass];
    self.atomicMassLabel.textAlignment = NSTextAlignmentRight;

    [self modifyLabel: self.boilingPointLabel withStandardConditionColor: self.element.standardConditionColor seriesColor: self.element.seriesColor];
    self.boilingPointLabel.text = [NSString stringWithFormat: @"%@", self.element.boilingPoint];
    self.boilingPointLabel.textAlignment = NSTextAlignmentRight;

    [self modifyLabel: self.meltingPointLabel withStandardConditionColor: self.element.standardConditionColor seriesColor: self.element.seriesColor];
    self.meltingPointLabel.text = [NSString stringWithFormat: @"%@", self.element.meltingPoint];
    self.meltingPointLabel.textAlignment = NSTextAlignmentRight;

    [self modifyLabel: self.densityLabel withStandardConditionColor: self.element.standardConditionColor seriesColor: self.element.seriesColor];
    self.densityLabel.text = [NSString stringWithFormat: @"%@", self.element.density];

    [self modifyLabel: self.seriesLabel withStandardConditionColor: self.element.standardConditionColor seriesColor: self.element.seriesColor];
    self.seriesLabel.text = [NSString stringWithFormat: @"%@", self.element.series];
    self.seriesLabel.textAlignment = NSTextAlignmentLeft;

    [self modifyLabel: self.periodLabel withStandardConditionColor: self.element.standardConditionColor seriesColor: self.element.seriesColor];
    self.periodLabel.text = [NSString stringWithFormat: @"%@", self.element.period];
    self.periodLabel.textAlignment = NSTextAlignmentRight;

    [self modifyLabel: self.groupLabel withStandardConditionColor: self.element.standardConditionColor seriesColor: self.element.seriesColor];
    self.groupLabel.text = [NSString stringWithFormat: @"%@", self.element.group];
    self.groupLabel.textAlignment = NSTextAlignmentRight;
}

- (void)dealloc {
    [[NSNotificationCenter defaultCenter] removeObserver: self name: NOTIFICATION_SERIES_COLOR_CHANGED object: nil];
}

@end

File: PrimaMateria/PrimaMateria/Classes/XTRElementTableViewCell.swift
//
//  XTRElementTableViewCell.swift
//  PrimaMateria
//
//  Created by Jerry Porter on 4/15/16.
//  Copyright ©2018 xTrensa. All rights reserved.
//

class XTRElementTableViewCell : UITableViewCell {
    
    var atomicNumberLabel : XTRTableLabel!
    var symbolLabel : XTRTableLabel!
    var nameLabel : XTRTableLabel!
    var atomicMassLabel : XTRTableLabel!
    var boilingPointLabel : XTRTableLabel!
    var meltingPointLabel : XTRTableLabel!
    var densityLabel : XTRTableLabel!
    var seriesLabel : XTRTableLabel!
    var periodLabel : XTRTableLabel!
    var groupLabel : XTRTableLabel!
    var element : XTRElement!
    
    // MARK: - Initialization Methods
    
    required override init(style: UITableViewCellStyle, reuseIdentifier: String?) {
        super.init(style: style, reuseIdentifier: reuseIdentifier)
    }
    
    func setupWithElement(_ anElement: XTRElement) {
        element = anElement
        backgroundColor = UIColor.black
        contentView.backgroundColor = UIColor.darkGray
        
        atomicNumberLabel = XTRTableLabel(frame: CGRect(x: 0, y: 1, width: 85, height: 42), backgroundColor: element.seriesColor)
        contentView.addSubview(atomicNumberLabel)
        
        symbolLabel = XTRTableLabel(frame: CGRect(x: 86, y: 1, width: 95, height: 42), backgroundColor: element.seriesColor)
        contentView.addSubview(symbolLabel)
        
        nameLabel = XTRTableLabel(frame: CGRect(x: 182, y: 1, width: 125, height: 42), backgroundColor: element.seriesColor)
        contentView.addSubview(nameLabel)
        
        atomicMassLabel = XTRTableLabel(frame: CGRect(x: 308, y: 1, width: 105, height: 42), backgroundColor: element.seriesColor)
        contentView.addSubview(atomicMassLabel)
        
        boilingPointLabel = XTRTableLabel(frame: CGRect(x: 414, y: 1, width: 105, height: 42), backgroundColor: element.seriesColor)
        contentView.addSubview(boilingPointLabel)
        
        meltingPointLabel = XTRTableLabel(frame: CGRect(x: 520, y: 1, width: 105, height: 42), backgroundColor: element.seriesColor)
        contentView.addSubview(meltingPointLabel)
        
        densityLabel = XTRTableLabel(frame: CGRect(x: 626, y: 1, width: 105, height: 42), backgroundColor: element.seriesColor)
        contentView.addSubview(densityLabel)
        
        seriesLabel = XTRTableLabel(frame: CGRect(x: 732, y: 1, width: 150, height: 42), backgroundColor: element.seriesColor)
        contentView.addSubview(seriesLabel)
        
        periodLabel = XTRTableLabel(frame: CGRect(x: 883, y: 1, width: 72, height: 42), backgroundColor: element.seriesColor)
        contentView.addSubview(periodLabel)
        
        groupLabel = XTRTableLabel(frame: CGRect(x: 956, y: 1, width: 68, height: 42), backgroundColor: element.seriesColor)
        contentView.addSubview(groupLabel)
        
        modifyCellProperties()
        NotificationCenter.default.addObserver(self, selector: #selector(XTRElementTableViewCell.colorChanged(_:)), name: .seriesColorChangedNotification, object: nil)
    }
    
    required init?(coder aDecoder: NSCoder) {
        super.init(coder: aDecoder)
    }
    
    // MARK: - Notification Methods
    
    @objc func colorChanged(_ aNotification: Notification) {
        setupColors()
        setNeedsDisplay()
    }
    
    // MARK: - Internal Methods
    
    func modifyLabel(_ aLabel: XTRTableLabel, withStandardConditionColor: UIColor, seriesColor: UIColor) {
        aLabel.font = UIFont.systemFont(ofSize: 16.0)
        aLabel.textColor = withStandardConditionColor
        aLabel.backgroundColor = seriesColor
        aLabel.autoresizingMask = [UIViewAutoresizing.flexibleRightMargin, UIViewAutoresizing.flexibleHeight]
    }
    
    func modifyCellProperties() {
        let seriesColor = element.seriesColor
        let selectedColor = seriesColor.inverseColor()
        
        selectedBackgroundView = UIImageView(image: UIImage.imageFromColor(selectedColor, andSize: frame.size))
        
        modifyLabel(atomicNumberLabel, withStandardConditionColor: element.standardConditionColor, seriesColor: element.seriesColor)
        atomicNumberLabel.text = String(element.atomicNumber)
        atomicNumberLabel.textAlignment = .right
        
        modifyLabel(symbolLabel, withStandardConditionColor: element.standardConditionColor, seriesColor: element.seriesColor)
        symbolLabel.font = UIFont.systemFont(ofSize: 26.0)
        symbolLabel.text = String(format: "%@", element.symbol!)
        symbolLabel.textAlignment = .center
        
        modifyLabel(nameLabel, withStandardConditionColor: element.standardConditionColor, seriesColor: element.seriesColor)
        nameLabel.text = String(format: "%@", element.name!)
        nameLabel.textAlignment = .left
        
        modifyLabel(atomicMassLabel, withStandardConditionColor: element.standardConditionColor, seriesColor: element.seriesColor)
        atomicMassLabel.text = element.value(forKeyPath: ELEMENT_ATOMIC_MASS) as? String
        atomicMassLabel.textAlignment = .right
        
        modifyLabel(boilingPointLabel, withStandardConditionColor: element.standardConditionColor, seriesColor: element.seriesColor)
        boilingPointLabel.text = element.value(forKeyPath: ELEMENT_BOILING_POINT) as? String
        boilingPointLabel.textAlignment = .right
        
        modifyLabel(meltingPointLabel, withStandardConditionColor: element.standardConditionColor, seriesColor: element.seriesColor)
        meltingPointLabel.text = element.value(forKeyPath: ELEMENT_MELTING_POINT) as? String
        meltingPointLabel.textAlignment = .right
        
        modifyLabel(densityLabel, withStandardConditionColor: element.standardConditionColor, seriesColor: element.seriesColor)
        densityLabel.text = element.value(forKeyPath: ELEMENT_DENSITY) as? String
        densityLabel.textAlignment = .right
        
        modifyLabel(seriesLabel, withStandardConditionColor: element.standardConditionColor, seriesColor: element.seriesColor)
        seriesLabel.text = String(format: "%@", element.series!)
        seriesLabel.textAlignment = .left
        
        modifyLabel(periodLabel, withStandardConditionColor: element.standardConditionColor, seriesColor: element.seriesColor)
        periodLabel.text = String(format: "%@", element.period!)
        periodLabel.textAlignment = .right
        
        modifyLabel(groupLabel, withStandardConditionColor: element.standardConditionColor, seriesColor: element.seriesColor)

        let aGroup = element.group!
        groupLabel.text = (aGroup != 0) ? String(format: "%d", aGroup) : STRING_EMPTY
        groupLabel.textAlignment = .right
    }
    
    func setupColors() {
        let aSeriesColor = element.seriesColor
        
        atomicNumberLabel.backgroundColor = aSeriesColor
        symbolLabel.backgroundColor = aSeriesColor
        nameLabel.backgroundColor = aSeriesColor
        atomicMassLabel.backgroundColor = aSeriesColor
        boilingPointLabel.backgroundColor = aSeriesColor
        meltingPointLabel.backgroundColor = aSeriesColor
        densityLabel.backgroundColor = aSeriesColor
        seriesLabel.backgroundColor = aSeriesColor
        periodLabel.backgroundColor = aSeriesColor
        groupLabel.backgroundColor = aSeriesColor
    }
    
    // MARK: - Memory Management Methods
    
    deinit {
        NotificationCenter.default.removeObserver(self, name: .seriesColorChangedNotification, object: nil)
    }
    
}
